File: examples/boschProcess/boschProcessSimulate.py
from argparse import ArgumentParser
import viennaps as ps

# parse config file name and simulation dimension
parser = ArgumentParser(
    prog="boschProcess",
    description="Run a Bosch process simulation on a trench geometry.",
)
parser.add_argument("-D", "-DIM", dest="dim", type=int, default=2)
parser.add_argument("filename")
args = parser.parse_args()

# switch between 2D and 3D mode
if args.dim == 2:
    print("Running 2D simulation.")
else:
    print("Running 3D simulation.")
ps.setDimension(args.dim)

ps.Logger.setLogLevel(ps.LogLevel.ERROR)
params = ps.readConfigFile(args.filename)
ps.setNumThreads(16)

geometry = ps.Domain(
    gridDelta=params["gridDelta"],
    xExtent=params["xExtent"],
    yExtent=params["yExtent"],
)
ps.MakeTrench(
    domain=geometry,
    trenchWidth=params["trenchWidth"],
    trenchDepth=0.0,
    maskHeight=params["maskHeight"],
).apply()

# Isotropic deposition model
depoModel = ps.SingleParticleProcess(
    rate=params["depositionThickness"],
    stickingProbability=params["depositionStickingProbability"],
)

# Deposition removal model
depoRemoval = ps.SingleParticleProcess(
    rate=-params["depositionThickness"],
    stickingProbability=1.0,
    sourceExponent=params["ionSourceExponent"],
    maskMaterial=ps.Material.Mask,
)

# Etch model
etchModel = ps.MultiParticleProcess()
etchModel.addNeutralParticle(params["neutralStickingProbability"])
etchModel.addIonParticle(sourcePower=params["ionSourceExponent"], thetaRMin=60.0)


# Custom rate function for the etch model
def rateFunction(fluxes, material):
    if material == ps.Material.Mask:
        return 0.0
    rate = fluxes[1] * params["ionRate"]
    if material == ps.Material.Si:
        rate += fluxes[0] * params["neutralRate"]
    return rate


etchModel.setRateFunction(rateFunction)
etchTime = params["etchTime"]

n = 0


def runProcess(model, name, time=1.0):
    global n
    print("  - {} - ".format(name))
    ps.Process(geometry, model, time).apply()
    geometry.saveSurfaceMesh("boschProcessSimulate_{}".format(n))
    n += 1


numCycles = int(params["numCycles"])

# Initial geometry
geometry.saveSurfaceMesh("boschProcessSimulate_{}".format(n))
n += 1

runProcess(etchModel, "Etching", etchTime)

for i in range(numCycles):
    print("Cycle {}".format(i + 1))

    # Deposit a layer of polymer
    geometry.duplicateTopLevelSet(ps.Material.Polymer)
    runProcess(depoModel, "Deposition")

    # Remove the polymer layer
    runProcess(depoRemoval, "Punching through")

    # Etch the trench
    runProcess(etchModel, "Etching", etchTime)

    # Ash (remove) the polymer
    geometry.removeTopLevelSet()
    geometry.removeStrayPoints()
    geometry.saveSurfaceMesh("boschProcessSimulate_{}".format(n))
    n += 1

# save the final geometry
geometry.saveVolumeMesh("boschProcessSimulate_final")

File: examples/boschProcess/boschProcessSimulate.cpp
#include <geometries/psMakeTrench.hpp>
#include <models/psIsotropicProcess.hpp>
#include <models/psMultiParticleProcess.hpp>
#include <models/psSingleParticleProcess.hpp>
#include <process/psProcess.hpp>
#include <psUtil.hpp>

using namespace viennaps;
constexpr int D = 2;
using NumericType = double;
const std::string name = "boschProcessSimulate_";

void etch(SmartPointer<Domain<NumericType, D>> domain, util::Parameters &params,
          int &n) {
  std::cout << "  - Etching - " << std::endl;
  auto etchModel = SmartPointer<MultiParticleProcess<NumericType, D>>::New();
  etchModel->addNeutralParticle(params.get("neutralStickingProbability"));
  etchModel->addIonParticle(params.get("ionSourceExponent"),
                            60.0 /*thetaRMin*/);
  const NumericType neutralRate = params.get("neutralRate");
  const NumericType ionRate = params.get("ionRate");
  etchModel->setRateFunction(
      [neutralRate, ionRate](const std::vector<NumericType> &fluxes,
                             const Material &material) {
        if (material == Material::Mask)
          return 0.;
        NumericType rate = fluxes[1] * ionRate;
        if (material == Material::Si)
          rate += fluxes[0] * neutralRate;
        return rate;
      });
  Process<NumericType, D>(domain, etchModel, params.get("etchTime")).apply();
  domain->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
}

void punchThrough(SmartPointer<Domain<NumericType, D>> domain,
                  util::Parameters &params, int &n) {
  std::cout << "  - Punching through - " << std::endl;
  NumericType depositionThickness = params.get("depositionThickness");

  // punch through step
  auto depoRemoval = SmartPointer<SingleParticleProcess<NumericType, D>>::New(
      -depositionThickness, 1. /* sticking */, params.get("ionSourceExponent"),
      Material::Mask);
  Process<NumericType, D>(domain, depoRemoval, 1.).apply();
  domain->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
}

void deposit(SmartPointer<Domain<NumericType, D>> domain,
             util::Parameters &params, int &n) {
  std::cout << "  - Deposition - " << std::endl;
  NumericType depositionThickness = params.get("depositionThickness");
  NumericType depositionSticking = params.get("depositionStickingProbability");
  domain->duplicateTopLevelSet(Material::Polymer);
  auto model = SmartPointer<SingleParticleProcess<NumericType, D>>::New(
      depositionThickness, depositionSticking);
  Process<NumericType, D>(domain, model, 1.).apply();
  domain->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
}

void ash(SmartPointer<Domain<NumericType, D>> domain, int &n) {
  domain->removeTopLevelSet();
  domain->removeStrayPoints();
  domain->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
}

int main(int argc, char **argv) {

  Logger::setLogLevel(LogLevel::ERROR);
  omp_set_num_threads(16);

  // Parse the parameters
  util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    // Try default config file
    params.readConfigFile("config.txt");
    if (params.m.empty()) {
      std::cout << "No configuration file provided!" << std::endl;
      std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
      return 1;
    }
  }

  // geometry setup
  auto geometry = Domain<NumericType, D>::New(
      params.get("gridDelta"), params.get("xExtent"), params.get("yExtent"));
  MakeTrench<NumericType, D>(geometry, params.get("trenchWidth"),
                             0.0, // trenchDepth
                             0.0, // trenchTaperAngle
                             params.get("maskHeight"))
      .apply();

  const int numCycles = params.get<int>("numCycles");

  int n = 0;
  geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
  etch(geometry, params, n);

  for (int i = 0; i < numCycles; ++i) {
    std::cout << "Cycle " << i + 1 << std::endl;
    deposit(geometry, params, n);
    punchThrough(geometry, params, n);
    etch(geometry, params, n);
    ash(geometry, n);
  }

  geometry->saveVolumeMesh(name + "final");
}